feat(helm): Allow configurable severity and thresholds for alerts

schahal opened the pull request:
**What this PR does / why we need it**:

The Loki helm chart comes with hardcoded `PrometheusRules` where users are unable to change the e.g., severity or threshold to best fit their environment's behavior.

This PR allows users to set some overrides.

For example, in my environment, a `critical` severity may mean a phone call in the middle of the night. So I have a values.yaml with the following override:

```
monitoring:
  rules:
    configs:
      LokiRequestPanics:
        # warn me if there are more than 3 panics over the last hour
        lookbackPeriod: 60m
        severity: warning
        threshold: 2
```

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:

* This is simply pulling the default attributes of the PrometheusRule expressions into `values.yaml`
* This is _**backwards compatible**_ with the `monitoring.rules.disabled` keys (marking that as deprecated and to use configs for enabling/disabling). But we keep the deprecated keys for backwards compatibility
* **Testing**: Ran `helm template test --set loki.useTestSchema=true --set monitoring.rules.enabled=true --debug . --validate` and confirmed not rendered diff before and after.

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [x] Title matches the required conventional commits format, see [here](https://www.conventionalcommits.org/en/v1.0.0/)
  - **Note** that Promtail is considered to be feature complete, and future development for logs collection will be in [Grafana Alloy](https://github.com/grafana/alloy). As such, `feat` PRs are unlikely to be accepted unless a case can be made for the feature actually being a bug fix to existing behavior.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [x] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/loki?pullRequest=13730) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/loki?pullRequest=13730) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/loki?pullRequest=13730) it.</sub>

schahal commented:
Hi @JStickler @DylanGuedes @chaudum @vlad-diachenko  : I'm wondering how we can get a reviewer/approval for this? I don't see any guidance in the Contributors doc or guidelines.

Tagged you as I saw the last couple `feat(helm)` approvals came from you (apologies in advance if it should be someone else, or if I missed something in `CONTRIBUTING.md`).

This is to address/resolve https://github.com/grafana/loki/issues/11219

schahal commented:
Adding @trevorwhitney and @paul1r also for approval of the CI workflows and possible review/approve/merge?

schahal commented:
Hi @jkroepke , yeah I think this is still needed (as I've simply worked around it by disabling the rules altogether).

I've updated the branch to (a) remove the Chart version bump, (b) updated the CHANGELOG, (c) `additionalRuleAnnotations` has been added since I last worked on this branch, updated its rendering.

schahal commented on production/helm/loki/src/alerts.yaml.tpl:
```diff
@@ -2,96 +2,110 @@
 {{- end }}
 
-{{- if not (.Values.monitoring.rules.disabled.LokiCanaryLatency | default false) }}
+{{- if not (.disabled.LokiCanaryLatency | default false) }}
```
Whoops missed this during merge-conflict-resolutions (this branch was opened over a year ago and this was added since then).

Updated in latest PR.

schahal commented on production/helm/loki/src/alerts.yaml.tpl:
```diff
@@ -2,96 +2,110 @@
+{{- with .Values.monitoring.rules }}
+{{- $additionalAnnotations := .additionalRuleAnnotations }}
+{{- $additionalLabels := .additionalRuleLabels }}
+{{- if not (.disabled.LokiRequestErrors | default (not .configs.LokiRequestErrors.enabled)) }}
```
Updated in last commit. Good catch!!

`and()` is safer and cleaner to read in fact.

Tested by setting `disabled.LokiRequestErrors` to true and diff'd rendered before/after:

```diff
1515,1528d1514
<         - alert: LokiRequestErrors
<           annotations:
<             message: |
<               {{ $labels.job }} {{ $labels.route }} is experiencing {{ printf "%.2f" $value }}% errors.
<             shum: fa
<           expr: |
<             100 * sum(rate(loki_request_duration_seconds_count{status_code=~"5.."}[2m])) by (namespace, job, route)
<               /
<             sum(rate(loki_request_duration_seconds_count[2m])) by (namespace, job, route)
<               > 10
<           for: 15m
<           labels:
<             severity: critical
```

Jayclifford345 reviewed (approved):
LGTM
